src/traversal.rs

```diff
@@ -0,0 +1,134 @@
+array_of_tables: Type representing a TOML array of tables
+item:            Type representing either a value, a table, an array of tables, or none.
+table:           Type representing a TOML non-inline table
+value:           Representation of a TOML Value (as part of a Key/Value Pair).
+
+*/
+
+pub fn traverse<'a>(field: &str, doc: &'a mut DocumentMut, op: TraverseOps) -> Result<Json> {
+    let path_split = field.split('/').collect::<Vec<&str>>();
+    let path_slice = path_split.as_slice();
+    let root_key = path_split.get(0).ok_or(anyhow!("Invalid query path!"))?;
+    let table = doc.as_table_mut();
+    let item = table
+        .get_mut(root_key)
+        .ok_or(anyhow!("Missing table for traversal"))?;
+    do_traverse(&path_slice[1..], &mut At::Item::<'a>(item), op)
+}
+
+fn do_traverse(path: &[&str], item: &mut At, op: TraverseOps) -> Result<Json> {
+    if !path.is_empty() {
+        return item.fold(path[0], &path[1..], op);
+    }
+
+    match op {
+        TraverseOps::Get(callback) => callback(item),
+    }
+}
+
+impl At<'_> {
+    fn fold(&mut self, key: &str, path: &[&str], op: TraverseOps) -> Result<Json> {
+        match self {
+            At::Array(arr) => {
+                let index = key
+                    .parse::<usize>()
+                    .map_err(|_| anyhow!("Key is not a valid usize"))?;
+                let member = arr.get_mut(index).ok_or(anyhow!("Array out of range"))?;
+                do_traverse(path, &mut At::Value(member), op)
+            }
+            At::ArrayOfTables(aar) => {
+                let index = key
+                    .parse::<usize>()
+                    .map_err(|_| anyhow!("Key is not a valid usize"))?;
+                let member = aar.get_mut(index).ok_or(anyhow!("Array out of range"))?;
+                do_traverse(path, &mut At::Table(member), op)
+            }
+            At::Item(item) => match item {
+                Item::ArrayOfTables(aar) => At::ArrayOfTables(aar).fold(key, path, op),
+                Item::Table(table) => At::Table(table).fold(key, path, op),
+                Item::Value(value) => At::Value(value).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index item {:?} with {:?}", item, key)),
+            },
+            At::Table(table) => {
+                let mut found = table
+                    .get_mut(key)
+                    .ok_or(anyhow!("Unable to index table with {:?}", key))?;
+                do_traverse(path, &mut At::Item(&mut found), op)
+            }
+            At::Value(value) => match value {
+                Value::Array(arr) => At::Array(arr).fold(key, path, op),
+                _ => Err(anyhow!("Unable to index value {:?} with {:?}", value, key)),
+            },
+        }
+    }
+
+    pub fn to_value(&mut self) -> Result<serde_json::Value> {
+        match self {
+            At::Array(_arr) => Err(anyhow!("array")),
+            At::ArrayOfTables(_aar) => Err(anyhow!("stub")),
```

Comment:
```suggestion
            At::ArrayOfTables(_aar) => Err(anyhow!("stub array of tables")),
```

Reply:
Actually missed this in the first pass, pushing up the impl now. There should be no unimplemented branches.